File: backend/app/chunking.py
"""Split page text into overlapping chunks while tracking exact character spans.

Invariant: for every chunk, ``page_text[chunk.start:chunk.end] == chunk.text``.
That guarantee is what lets the frontend highlight the exact source span of a
citation, so chunk boundaries are always cut from the original string - the
text is never rewritten, joined, or re-whitespaced.
"""

import re
from dataclasses import dataclass

from . import config

_PARA_BREAK = re.compile(r"\n{2,}")
_SENTENCE_BREAK = re.compile(r"(?<=[.!?])\s+")

Span = tuple[int, int]


@dataclass(frozen=True)
class Chunk:
    chunk_id: str
    doc_id: str
    page: int  # 1-based
    start: int  # offsets into the page's text
    end: int
    text: str


def _strip_span(text: str, start: int, end: int) -> Span | None:
    while start < end and text[start].isspace():
        start += 1
    while end > start and text[end - 1].isspace():
        end -= 1
    return (start, end) if end > start else None


def _paragraph_spans(text: str) -> list[Span]:
    spans: list[Span] = []
    pos = 0
    for match in _PARA_BREAK.finditer(text):
        if (span := _strip_span(text, pos, match.start())) is not None:
            spans.append(span)
        pos = match.end()
    if (span := _strip_span(text, pos, len(text))) is not None:
        spans.append(span)
    return spans


def _sentence_spans(text: str, start: int, end: int, hard_max: int) -> list[Span]:
    """Split an oversized paragraph on sentence boundaries (hard-cut as a last resort)."""
    sentences: list[Span] = []
    pos = start
    for match in _SENTENCE_BREAK.finditer(text, start, end):
        if (span := _strip_span(text, pos, match.start())) is not None:
            sentences.append(span)
        pos = match.end()
    if (span := _strip_span(text, pos, end)) is not None:
        sentences.append(span)

    result: list[Span] = []
    for s, e in sentences:
        while e - s > hard_max:
            result.append((s, s + hard_max))
            s += hard_max
        result.append((s, e))
    return result


def chunk_page(
    text: str,
    *,
    doc_id: str,
    page: int,
    target: int = config.CHUNK_TARGET_CHARS,
    overlap: int = config.CHUNK_OVERLAP_CHARS,
    hard_max: int = config.CHUNK_HARD_MAX_CHARS,
) -> list[Chunk]:
    """Greedily pack paragraphs into ~target-char chunks with ~overlap-char overlap."""
    units: list[Span] = []
    for s, e in _paragraph_spans(text):
        if e - s <= hard_max:
            units.append((s, e))
        else:
            units.extend(_sentence_spans(text, s, e, hard_max))

    if not units:
        return []

    spans: list[Span] = []
    i = 0
    while i < len(units):
        j = i
        while j + 1 < len(units) and units[j + 1][1] - units[i][0] <= target:
            j += 1
        chunk_start, chunk_end = units[i][0], units[j][1]
        spans.append((chunk_start, chunk_end))
        if j + 1 >= len(units):
            break
        # Start the next chunk at the earliest unit that keeps the overlap
        # within budget; always advance past unit i so we make progress.
        k = j + 1
        while k - 1 > i and chunk_end - units[k - 1][0] <= overlap:
            k -= 1
        i = k

    return [
        Chunk(
            chunk_id=f"{doc_id}:{page}:{idx}",
            doc_id=doc_id,
            page=page,
            start=s,
            end=e,
            text=text[s:e],
        )
        for idx, (s, e) in enumerate(spans)
    ]


def chunk_document(doc_id: str, pages: list[str], **kwargs) -> list[Chunk]:
    chunks: list[Chunk] = []
    for page_index, text in enumerate(pages, start=1):
        chunks.extend(chunk_page(text, doc_id=doc_id, page=page_index, **kwargs))
    return chunks

File: frontend/src/lib/clause/chunking.ts
/** Split page text into overlapping chunks while tracking exact character spans.
 *
 * Invariant: for every chunk, pageText.slice(chunk.start, chunk.end) === chunk.text.
 * That guarantee is what lets the viewer highlight the exact source span of a
 * citation, so chunk boundaries are always cut from the original string. The
 * text is never rewritten, joined, or re-whitespaced.
 */

import type { Chunk } from "./types";

export const CHUNK_TARGET_CHARS = 1100;
export const CHUNK_OVERLAP_CHARS = 180;
export const CHUNK_HARD_MAX_CHARS = 2200;

/** [start, end) offsets into a page's text. */
type Span = [number, number];

const PARA_BREAK = /\n{2,}/g;
// Whitespace that follows sentence punctuation. Built fresh where it is used
// with lastIndex, so no shared scan position can leak between calls.
const SENTENCE_BREAK_SOURCE = "(?<=[.!?])\\s+";
const WHITESPACE = /\s/;

export interface ChunkOptions {
  target?: number;
  overlap?: number;
  hardMax?: number;
}

export interface ChunkPageOptions extends ChunkOptions {
  docId: string;
  page: number;
}

function isSpaceAt(text: string, index: number): boolean {
  const ch = text[index];
  return ch !== undefined && WHITESPACE.test(ch);
}

function stripSpan(text: string, start: number, end: number): Span | null {
  let s = start;
  let e = end;
  while (s < e && isSpaceAt(text, s)) s += 1;
  while (e > s && isSpaceAt(text, e - 1)) e -= 1;
  return e > s ? [s, e] : null;
}

function paragraphSpans(text: string): Span[] {
  const spans: Span[] = [];
  let pos = 0;
  for (const match of text.matchAll(PARA_BREAK)) {
    const span = stripSpan(text, pos, match.index);
    if (span !== null) spans.push(span);
    pos = match.index + match[0].length;
  }
  const tail = stripSpan(text, pos, text.length);
  if (tail !== null) spans.push(tail);
  return spans;
}

/** Split an oversized paragraph on sentence boundaries, hard-cutting as a last
 * resort so a wall of text with no punctuation still comes apart. */
function sentenceSpans(
  text: string,
  start: number,
  end: number,
  hardMax: number,
): Span[] {
  const sentences: Span[] = [];
  const breaks = new RegExp(SENTENCE_BREAK_SOURCE, "g");
  breaks.lastIndex = start;
  let pos = start;
  let match = breaks.exec(text);
  while (match !== null) {
    if (match.index >= end) break;
    const span = stripSpan(text, pos, match.index);
    if (span !== null) sentences.push(span);
    // The search region stops at end, so a run of whitespace never carries the
    // next unit's first character with it.
    pos = Math.min(match.index + match[0].length, end);
    match = breaks.exec(text);
  }
  const tail = stripSpan(text, pos, end);
  if (tail !== null) sentences.push(tail);

  const result: Span[] = [];
  for (const [sentenceStart, sentenceEnd] of sentences) {
    let s = sentenceStart;
    while (sentenceEnd - s > hardMax) {
      result.push([s, s + hardMax]);
      s += hardMax;
    }
    result.push([s, sentenceEnd]);
  }
  return result;
}

/** Greedily pack paragraphs into roughly target-sized chunks that overlap by
 * roughly overlap characters. */
export function chunkPage(text: string, options: ChunkPageOptions): Chunk[] {
  const {
    docId,
    page,
    target = CHUNK_TARGET_CHARS,
    overlap = CHUNK_OVERLAP_CHARS,
    hardMax = CHUNK_HARD_MAX_CHARS,
  } = options;

  const units: Span[] = [];
  for (const [s, e] of paragraphSpans(text)) {
    if (e - s <= hardMax) {
      units.push([s, e]);
    } else {
      units.push(...sentenceSpans(text, s, e, hardMax));
    }
  }
  if (units.length === 0) return [];

  const spans: Span[] = [];
  let i = 0;
  while (i < units.length) {
    const first = units[i];
    if (first === undefined) break;

    let j = i;
    for (;;) {
      const next = units[j + 1];
      if (next === undefined || next[1] - first[0] > target) break;
      j += 1;
    }
    const last = units[j];
    if (last === undefined) break;

    const chunkStart = first[0];
    const chunkEnd = last[1];
    spans.push([chunkStart, chunkEnd]);
    if (j + 1 >= units.length) break;

    // Start the next chunk at the earliest unit that keeps the overlap within
    // budget. Never walk back past unit i, so the loop always moves forward.
    let k = j + 1;
    for (;;) {
      if (k - 1 <= i) break;
      const prev = units[k - 1];
      if (prev === undefined || chunkEnd - prev[0] > overlap) break;
      k -= 1;
    }
    i = k;
  }

  return spans.map(([s, e], index) => ({
    chunkId: `${docId}:${page}:${index}`,
    docId,
    page,
    start: s,
    end: e,
    text: text.slice(s, e),
  }));
}

export function chunkDocument(
  docId: string,
  pages: string[],
  options: ChunkOptions = {},
): Chunk[] {
  const chunks: Chunk[] = [];
  pages.forEach((text, pageIndex) => {
    chunks.push(...chunkPage(text, { ...options, docId, page: pageIndex + 1 }));
  });
  return chunks;
}
